### ISSUE TITLE
HTTP3

### ISSUE THREAD

**Issue by alexhultman-2fa** [uNetworking/uWebSockets#1280]
As a prerequisite to WebTransport (over HTTP3) we need to support HTTP3. This was not initially a plan but since we do want to support the benefits of QUIC we need this. Lsquic has been benchmarked with great results (just as good as uSockets with TLS) which is a great indication of quality.

As a first step, the most basic and simple HTTP3 req/res needs to be supported so that a full benchmark can be done. Minimal or no API changes should be needed

**Comment by superdolt** [comment]
when do u think this will be implemented / done?

**Comment by superdolt** [comment]
by the way, is this also going to be the fastest http3 in the world?

**Comment by alexhultman** [comment]
It doesn't have to be, it just needs to be "significantly" faster than Node.js. And that is what I'm waiting for, they rely on OpenSSL 3.1 which is not released or planned yet, so I cannot really benchmark a full solution yet. I do not want to build something that has no motivation for existing, so it has to be significantly faster than Node.js. Luckily it seems that is the case for QUIC at least, so should be true for HTTP3 also. But it relies on lsquic.

**Comment by superdolt** [comment]
why not integrate with lsquic instead of developing your own?

**Comment by alexhultman** [comment]
Chrome 97 apparently has WebTransport..

**Comment by e3dio** [comment]
The problem is there are no servers with QUIC/WebTransport or nothing easy to use, this library needs to get that going

**Comment by alexhultman** [comment]
Just waiting for it to be standardized https://datatracker.ietf.org/doc/html/draft-ietf-webtrans-http3

**Comment by e3dio** [comment]
I am proud to say I am an official contributor to that Spec https://github.com/w3c/webtransport/commit/4930bfd0e0d61e438d67b408cd9ff2b93039058c

Although it appears my contribution has since been completely overwritten, so my 15 minutes of fame are over

**Comment by AmirHmZz** [comment]
It is standardized now !

**Comment by alexhultman** [comment]
Are you sure? Link?

**Comment by AmirHmZz** [comment]
> Are you sure? Link?

Oh, no I just wondered that it was renamed from QUIC to h3 officially, but it is still in draft stage.
I'm going to delete my comments after you see them.
Sorry again.

**Comment by ddevienne** [comment]
Hi Alex.  Just read in your _Capitalism_ issue that you are hard at work on WebTransport over HTTP/3.  Can you share what the client will be for testing?  And whether you are considering adding your own client, for testing, if existing clients are lacking?  Just curious.

**Comment by alexhultman** [comment]
I'm testing with all clients I can find - curl, quiche, lsquic, Firefox, Chrome

**Comment by hiqsociety** [comment]
these benchmark may help:
https://interop.seemann.io/

**Comment by alexhultman** [comment]
Most if not all of those "multi-benchmarking sites" such as TechEmpower and the like cannot even figure out how to benchmark a silly HTTP1 server properly without major contamination - so expecting someone to manage a proper benchmark of 15 low level quic implementations that each take at least 2 months of effort to even get rolling, is a fantasy. You might just as well stare at the sun for 45 minutes and go blind, then look inside your own mind and come up with more accurate numbers than those.

**Comment by alexhultman** [comment]
Here are some tentative perf. numbers:

* For reference, the following Node.js built-in non-SSL server does 16k req/sec on this machine:
```
const http = require('http');

const requestListener = function (req, res) {
  res.writeHead(200);
  res.end('Hello, World!');
}

const server = http.createServer(requestListener);
server.listen(8080);
```

* Same machine gets 200k SSL minimal req/seq with uSockets (nothing new here; faked http1 server without parser or formatter)
* For proper HTTP3 I currently get 84k req/sec if I use multiple QUIC connections but only one stream (this number is clearly better than anything Node.js but still way worse than the 200k)
* However, if we open two streams per connection we get 142k req/sec
* If we open 4 streams we get 250k req/sec (so at 4 streams we are above 200k)
* If we open 8 streams we get 345k req/sec
* If we open 16 streams per HTTP3 connection we get 475k req/sec (which is a bit ridiculous)

This is a bit apples-oranges comparison since HTTP3 streams are the equivalence to HTTP1 pipelining (which is considered broken and disabled in HTTP1 by all major browsers).

**Comment by alexhultman** [comment]
About 800k req/sec single-thread on M1 MacBook Air. That's with 16 streams and 10 connections.

**Comment by AmirHmZz** [comment]
**HTTP/3** is now _proposed as a standard_ by [RFC](https://datatracker.ietf.org/doc/rfc9204/). Congratulations to the web community :D

**Comment by theseyan** [comment]
Hi Alex, any updates on the progress of supporting WebTransports? I feel like there's huge potential in exposing a WebSocket like interface through a QUIC server and using WebTransports to communicate with it (as an alternative to Websockets in supported browsers)

**Comment by CMCDragonkai** [comment]
If you add support for quic, then does that mean this library would provide node devs a quic library even though node core disabled quic?

**Comment by alexhultman** [comment]
Yes uWS does not rely on anything nodejs, all it relies on is Google V8 for integration

**Comment by Zubnix** [comment]
Any plans on having websocket run over http3? https://datatracker.ietf.org/doc/rfc9220/

**Comment by Bolderaysky** [comment]
+1 for http3 support. While it isn't really as straightforward and easy to implement as http1.1, this would really improve performance and add a lot of new use cases for uwebsockets, where now http1.1 itself is performance/ bottleneck. Also, considering extensions, such as websockets over http3, it would make sense. Not really sure if this is already standard, but no doubt it will be really cool

**Comment by uNetworkingAB** [comment]
Yeah we are here for the long run and will eventually come back to HTTP3 at some point. No updates to post right now.

**Comment by sjmueller** [comment]
With `WebTransport` coming to Safari `iOS 18` and `macOS` this fall, accessibility could approach 85% of all users by end of 2024. Any update on if/when `uWebSockets` will add support?

**Comment by uNetworkingAB** [comment]
I cannot give you an exact update due to NDAs and other legal reasons, but I can say that the commercial interest in WebTransport in the fields I care about has been entirely missing. This is why development has stopped.

I don't want to be first to market just for the sake of being first to market - in fact, for WebSockets this library came 5 years after all the other implementations already had established their market shares.

If there is no proven commercial need for a feature to exist, it won't exist. As it would break the rules of free market capitalism.

If WebTransport becomes widespread and poorly implemented in common libraries, and where there is a significant commercial return on investment for implementing it better - I will do so.

Otherwise, I will not do so.

**Comment by sjmueller** [comment]
WebTransport already has a proven need in the game industry, as evidenced by most native mobile games which use udp. The web doesn’t have this option, except for WebRTC — which is a hacky solution for server authoritative games. WebTransport is the answer. 

Makes sense that you’re driven by your sponsorships + areas of interest, but “no proven commercial need” is confusing. Unless gaming isn’t considered commercial?

**Comment by uNetworkingAB** [comment]
Gaming can have a need for this, but the most important section still stands:

> If WebTransport becomes widespread and poorly implemented in common libraries, and where there is a significant commercial return on investment for implementing it better - I will do so.

So to speak clearly: give me a significant chunk of money and I will do it. Else, I will not do it.

**Comment by sjmueller** [comment]
> So to speak clearly: give me a significant chunk of money and I will do it. Else, I will not do it.

This makes sense to me, everyone should get a return on their time investment. Might be worthwhile to setup sponsorships in a way that allows the contributions to lean towards the features needed. Not sure if this is possible, but it would open up smaller contributors instead of relying on whales.

**Comment by m50S79sM6SRNp8Jn** [comment]
> > So to speak clearly: give me a significant chunk of money and I will do it. Else, I will not do it.
> 
> This makes sense to me, everyone should get a return on their time investment. Might be worthwhile to setup sponsorships in a way that allows the contributions to lean towards the features needed. Not sure if this is possible, but it would open up smaller contributors instead of relying on whales.

Sounds like a good plan. Did it ever get actioned, where?

**Comment by porsager** [comment]
Sounds perfect ! Can you give a ballpark estimate where a price negotiation will start, or would you rather do that privately business to business?

**Comment by m50S79sM6SRNp8Jn** [comment]
Anyone know if this is being taken a step further now?

**Comment by uNetworkingAB** [comment]
> Sounds perfect ! Can you give a ballpark estimate where a price negotiation will start, or would you rather do that privately business to business?

Start is 50k, that's definitely minimum

**Comment by uNetworkingAB** [comment]
> Anyone know if this is being taken a step further now?

> How's our *** AWESOME HTTP/3 QUICK WebTransport CODE FEATURE BOUNTY *** ?

You understand this issue is 4 years old already, right? It's not like something will happen on a daily basis here
